### PR TITLE
Fix ph5toms memory usage

### PR DESCRIPTION
**Pull request by derick-hess**
initial go at fixing memory usage. Removed yield in create cut and iterator in create_cut_list.
Looking at mprof I think there is still more that can be done

### CODE REVIEW

File: ph5/clients/ph5toms.py [hunk 1]
```diff
@@ -643,17 +661,25 @@ def create_cut_list(self):
...
+                            stations_list = self.create_cut(seed_network, ph5_station, seed_station,
+                                                   start_times, station_list, deployment, st_num)
+                            for station in stations_list:
+                                cuts_generator.append(station)
+
+        del self.ph5.Experiment_t
+        del self.ph5.Array_t
+        del self.ph5.Event_t
+        self.ph5.clear()
+        self.clear()
```
Comment (nick-iris) [inline ph5/clients/ph5toms.py]: Is it necessary to explicitly force the deletion of these? I seems like it should get handled automatically.

Comment (derick-hess) [inline ph5/clients/ph5toms.py]: with ph5.clear() they probably aren't necessary. I was deleting them before i started using ph5.clear() 

going back and looking at mprof plots it makes no difference.  We should get rid of those and just use ph5.clear()

ph5.clear() can make a decent difference of 10+MB depending on the experiment size

self.clear() makes a very very tiny difference. Less than a MB. Its basically just clearing all the request parameters after they are no longer needed but before the instance of the class is closed at the end of the program

**Comment by nick-iris** [comment]
I get the following error when running this:

```
(venv) falco-6:PH5 nick$ mprof run --include-children ph5toms --nickname master --ph5path /hdf5-data2/PH5_Experiments/pn4/16-015/ -o . --station 100? --channel DPZ --starttime 2016-06-26T00:00:00.0 --stoptime 2016-06-27T00:00:00.0
mprof: Sampling memory every 0.1s
Traceback (most recent call last):
  File "/Users/nick/Development/workspace/pic-ph5/venv/bin/ph5toms", line 11, in <module>
    load_entry_point('ph5', 'console_scripts', 'ph5toms')()
  File "/Users/nick/Development/workspace/PH5/ph5/clients/ph5toms.py", line 876, in main
    for t in streams:
  File "/Users/nick/Development/workspace/PH5/ph5/clients/ph5toms.py", line 677, in process_all
    cuts = self.create_cut_list()
  File "/Users/nick/Development/workspace/PH5/ph5/clients/ph5toms.py", line 666, in create_cut_list
    for station in stations_list:
TypeError: 'NoneType' object is not iterable
Closing remaining open files:/hdf5-data2/PH5_Experiments/pn4/16-015/master.ph5...done
```

**Comment by derick-hess** [comment]
Looks like a station isnt returning anything available. I'll fix that with check for None. That check was  able to just be done in the cut function when it was creating the obspy stream before

**Comment by nick-iris** [comment]
It should return a result though. See the following request:

https://service.iris.edu/ph5ws/station/1/query?network=YW&station=100?&channel=DPZ&starttime=2016-06-26T00:00:00.0&endtime=2016-06-27T00:00:00.0&nodata=404

This was the same example that I ran before when testing.

**Comment by derick-hess** [comment]
I'm not sure why a none type is returned I'll investigate that, but add a check for none type fixed it and the data is correct, but there is still a memory issue somewhere I need to work out. 
I made much a much bigger than the one you made but didnt tell it what channel I wanted and explicitly wrote out each station i wanted and the memory usage looked good, but the request you made the memory usage still continually goes up with the new code.

Sigh more investigation is needed to figure this one out

**Comment by nick-iris** [comment]
Ahh I think its because we were yielding in `create_cut` before which always returned an iterable. We need to change this function to return an empty list instead of None.

**Comment by derick-hess** [comment]
The memory usage on that request you made is the same if i request DPZ or all channels.
So that is interesting. It looks like it is still having issues garbage collecting

**Comment by derick-hess** [comment]
Okay a little more testing is required but I think I fixed it and possibly can add back in the nested yields.  If we can use the nested yields that would save a few seconds to the first file write out

**Comment by derick-hess** [comment]
After testing I can do both yields. The fix was way simpler than I thought.

**Comment by derick-hess** [comment]
This is still with one yield. I'm testing with 2 yields and if it works i'll push again

**Comment by nick-iris** [comment]
What test are you running? The following test still shows high memory usage:
```
mprof run --include-children ph5toms --nickname master --ph5path /hdf5-data2/PH5_Experiments/pn4/16-015/ -o . --station 100? --channel DPZ --starttime 2016-06-26T00:00:00.0 --stoptime 2016-06-27T00:00:00.0
```
screen shot 2017-08-18 at 11 42 22 am

**Comment by derick-hess** [comment]
```
mprof run --include-children ph5toms -n master.ph5 -p /mnt/Wavefields/PROCESS/Sigma/16-015/ -s 2016-06-26T00:00:00.00 -t 2016-06-27T00:00:00.0 --channel DPZ --station 100? -o ~/Derick_mseed/
```

I get this:

screen

**Comment by derick-hess** [comment]
I dont have a graph from before the changes. I can make one, but before the graph sould slowly ramp up to about 1400MB

**Comment by derick-hess** [comment]
Here is what the old code looks like on that same request

**Comment by derick-hess** [comment]
I was able to add both yields back in when I added both back in it didn't change overall memory usage. just shifted when the peak happens.

**Comment by derick-hess** [comment]
I added both yields back in as a test. It works and peak memory doesnt change. Time to first fiel output is a bit faster but overall execution on that request is 8 seconds longer.
Not sure if it is worth it to push the change that adds both yields back in.

**Comment by nick-iris** [comment]
I would say that the time to return a result is more important than the overall execution time (depending on how much time we are talking).

**Comment by derick-hess** [comment]
I'll go ahead and push the nested yield version. The difference on a small request like the one above is just under 1 second but on a really large request it could be multiple seconds so you are right we should probably go for time to first result

**Comment by derick-hess** [comment]
Just to test it on a different machine I tested it on ops.iris.washington.edu with the same parameters but also using the wavefields in /hdf-data2/PH5_Experiments/pn4/16-015

here is that graph:

**Comment by derick-hess** [comment]
no problem. There are probably more memory usage tweaks we can make in the future. For now this  makes sure we don't leak memory and cause a request to fail.

**Comment by nick-iris** [comment]
Do you want to create a new tagged release v4.0.2? I also fixed a slow method in the station service code that I was planning to roll out with this.

**Comment by derick-hess** [comment]
I just finished the ph5tostationxml issue #107 If we want to add that into 4.0.2 as well. I"m about to push that PR

**Pull request merged**